rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+For Aurora, it can be any of the exposed database endpoints, reader or writer.
+* `ca_path` is a path to the [RDS root certificate](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html).
+* `aws.auth` is the authentication method, only "iam" is supported.
+* `aws.region` is the region the database is deployed in.
+
+AWS credentials, if left unspecified, will be retrieved from the environment
+in the same way as `aws` CLI tool does, i.e. from the credentials file or
+supported environment variables. Credentials can also be provided in the 
+service configuration via `aws.access_key` and `aws.secret_key` fields.
+
+### Database instance
+
+#### Onprem
+
+To support mutual TLS auth, onprem database instance needs to be configured
+with Teleport's certificate authority and (optionally) cert/key pair issued
+by Teleport. The existing `tctl auth sign` command is used to produce these
+secrets:
+
+```sh
+$ tctl auth sign --format=db --host=localhost --out=server --ttl=8760h
+$ ls
+server.cas server.crt server.key
+```
+
+Note that the `--host` parameter should match the hostname of the endpoint
+the database will be connected at.
+
+The certificate is signed by Teleport host authority which is intended for 
+machine-to-machine communication.
+
+*TODO(r0mant): Instead of using host authority, introduce another authority
+specifically for db access? Would also help decouple authority rotation for
+databases from the rest of the cluster.*
+
+The generated secrets are then used to configure the database. For example,
+for PostgreSQL:
+
+```sh
+$ cat postgresql.conf | grep ssl
+ssl = on
+ssl_ca_file = '/path/to/server.cas'
+ssl_cert_file = '/path/to/server.crt'
+ssl_key_file = '/path/to/server.key'
+```
+
+In addition, PostgreSQL access configuration file should be configured to 
+require client certificate authentication:
+
+```sh
+$ cat pg_hba.conf
+# TYPE  DATABASE        USER            ADDRESS                 METHOD
+hostssl all             all             ::/0                    cert
+hostssl all             all             0.0.0.0/0               cert
+```
+
+#### RDS/Aurora
+
+There are a few things that need to be done in order to configure RDS/Aurora
+database to support IAM authentication.
+
+1. Enable IAM authentication when provisioning a new database, or on an
+existing one. By default, only password authentication is enabled.
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Enabling.html
+
+2. Configure IAM policy in order to allow the user Teleport database service
+will use to connect to the database instance with the auth token. An example
+policy is shown below.
+
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": "rds-db:connect",
+            "Resource": [
+                "arn:aws:rds-db:us-west-1:1234567890:dbuser:db-ABCDEFGHIJKLMNOP/developer",
+            ]
+        }
+    ]
+}
+```
```

Comment:
explain how teleport maps users to IAM roles using RBAC here, this is critical piece.

Reply:
Hmm, currently it doesn't really. The idea is that the "database service" is configured with the policy like shown here which allows IAM-based auth with specified databases as specified database users. So a Teleport user should first pass all Teleport's RBAC checks and then the db service connects to the database which is subject to this IAM check according to this defined policy.

We could use something like AWS impersonation API ("assume-role") but I'm not sure what it would get us, and would also require somehow getting/passing the role arn a user wants to assume from the user to the db service... Is there a use-case I'm not thinking of?

Reply:
this makes sense, I was asking how IAM enables mapping of teleport users to DB users.

Reply:
You explained in the comment below, please update RFD with clarification